RHDHPAI-698: manifests, and accompanying code changes, to bring up bridge as RHDH sidecar containers

gabemontero opened the pull request:
So this was interesting out of the gate because RHDH forces the backstage deployment to `automountServiceAccountToken: false`
meaning we have to manually mount the SA credentials/certs we want the bridge to use.

Ultimately, I could got the correct k8s json patch specification for the entire deployment such that it would add/merge the the new secret `rhdh-rhoai-bridge-token` volume with all the other volumes mounted as part of the ai-rhdh-installer into the deployment.

Instead, I import the secret via `envFrom` into our containers, and also add a volumeMount to the root volume declared for the deployment (i.e. /opt/app-root-src) and then store the ca.crt from the env var into a file which I can then seed into the REST config used to establish k8s client communications.

I also dynamically set the pod's IP and namespace as env vars as well.  This allows us to create our storage configmap as well as supply the root IP for the communication between the 4 containers in the pod.

Some more work will be needed to supply the static token for external access to the catalog to the backstage app config, as well as the allowed hosts.  A '10.*' entry may be needed.

Lastly, so far I did not need to create additional Services for the bridge's REST endpoints to prove the config map is populated when a new model is created/deployed from the model registry.  But as we complete integrating with our new backstage extensions, if need be that would need to be added as new instances in our assets/sidecar directory, using the same selector as the backstage service.

gabemontero commented:
yeah @johnmcollier @thepetk using the `?w=1` suffix on the files changed view to filter out white space

gabemontero commented on assets/sidecar/backstage-cr-patches.yaml:
```diff
@@ -0,0 +1,96 @@
+                envFrom:
+                  - secretRef:
+                      name: rhdh-rhoai-bridge-token
+                image: quay.io/gabemontero/model-catalog-bridge-location:latest
```
will switch to our redhat-ai-dev image after review from you all and some more testing wrt trying to get the app-config set up for allowed hosts and external tokens

gabemontero commented on assets/sidecar/backstage-cr-patches.yaml:
```diff
@@ -0,0 +1,96 @@
+                envFrom:
+                  - secretRef:
+                      name: rhdh-rhoai-bridge-token
+                image: quay.io/gabemontero/model-catalog-bridge-storage-rest:latest
```
will switch to our redhat-ai-dev image after review from you all and some more testing wrt trying to get the app-config set up for allowed hosts and external tokens

gabemontero commented on assets/sidecar/backstage-cr-patches.yaml:
```diff
@@ -0,0 +1,96 @@
+                envFrom:
+                  - secretRef:
+                      name: rhdh-rhoai-bridge-token
+                image: quay.io/gabemontero/model-catalog-bridge-rhoai-normalizer:latest
```
will switch to our redhat-ai-dev image after review from you all and some more testing wrt trying to get the app-config set up for allowed hosts and external tokens

gabemontero commented:
So far though I've been able to pull metadata for a model deployed from model registry and see the location svc and our storage configmap correctly populated @johnmcollier @thepetk

gabemontero commented:
so with the manifest changes I posted today @johnmcollier @thepetk when I manually patch the app-config config map with an actual token in the env var secret, I can query my RHDH catalog from the cli:

```
gmontero ~/go/src/github.com/redhat-ai-dev/model-catalog-bridge  (sidecar-cfg)$ bac get resources --backstage-url=https://backstage-ai-rh-developer-hub-ai-rhdh.apps.gmontero416.devcluster.openshift.com --backstage-skip-tls=true --backstage-token=$RHDH_TOKEN
I0331 17:17:21.090867  140011 utils.go:52] [
    {
        "metadata": {
            "uid": "9e058e53-e22d-4ef0-ab11-b226ca491677",
            "etag": "98edf3ec112af35f49d23e69778a0020181ea112",
            "name": "v1",
            "namespace": "default",
            "description": "A custom Red Hat base model instruct tuned only for phase 00, produced by IBM Research specifically for RHEL AI.",
            "annotations": {
                "backstage.io/edit-url": "https://github.com/gabemontero-appstudio-test-org/catalog-info-yaml/edit/master/catalog-info.yaml",
                "backstage.io/managed-by-location": "url:https://github.com/gabemontero-appstudio-test-org/catalog-info-yaml/tree/master/catalog-info.yaml",
                "backstage.io/managed-by-origin-location": "url:https://github.com/gabemontero-appstudio-test-org/catalog-info-yaml/blob/master/catalog-info.yaml",
                "backstage.io/source-location": "url:https://github.com/gabemontero-appstudio-test-org/catalog-info-yaml/tree/master/",
                "backstage.io/techdocs-ref": "resource/",
                "backstage.io/view-url": "https://github.com/gabemontero-appstudio-test-org/catalog-info-yaml/tree/master/catalog-info.yaml"
            },
            "links": [
                {
                    "url": "oci://registry.redhat.io/rhelai1/modelcar-granite-3-1-8b-starter-v1:1.4.0",
                    "icon": "WebAsset",
                    "type": "website"
                }
            ]
        },
        "relations": [
            {
                "type": "dependencyOf",
                "targetRef": "component:default/granite-3.1-8b-starter-v1-1.4.0",
                "target": {
                    "name": "",
                    "kind": "",
                    "namespace": ""
                }
            },
            {
                "type": "ownedBy",
                "targetRef": "user:default/kube:admin",
                "target": {
                    "name": "",
                    "kind": "",
                    "namespace": ""
                }
            }
        ],
        "apiVersion": "backstage.io/v1alpha1",
        "kind": "Resource",
        "spec": {
            "type": "ai-model",
            "lifecycle": "development",
            "owner": "user:kube:admin",
            "dependencyOf": [
                "component:granite-3.1-8b-starter-v1-1.4.0"
            ],
            "profile": {
                "displayName": "v1"
            }
        }
    }
]
gmontero ~/go/src/github.com/redhat-ai-dev/model-catalog-bridge  (sidecar-cfg)$ ```

So the static external access token is working correctly.

However, when the storage-rest container tries to import a location to backstage, it gets an error complaining about auth:

```
I0331 21:19:41.683546       1 rest.go:74] GGM postToBackstage url http://10.128.2.58:7007/api/catalog/locations token <the correct value for my static token that I verified with the CLI>
E0331 21:19:41.683594       1 server.go:294] error importing location http://10.128.2.58:9090/mnist/v1/catalog-info.yaml to backstage: Post "http://10.128.2.58:7007/api/catalog/locations": net/http: invalid header field value for "Authorization"
```

Maybe that means my allowed host setting of `10.128.*` is not valid for allowing the import, and we have to define a service ? .... didn't see anything useful in the backstage logs, but will pick up on Tuesday.

gabemontero commented on assets/sidecar/developer-hub-app-config-patches.yaml:
```diff
@@ -0,0 +1,113 @@
+          production:
+            clientId: ${GITHUB__APP__CLIENT__ID}
+            clientSecret: ${GITHUB__APP__CLIENT__SECRET}
+    # start of delta from ai-rhdh-installer
```
start of the changes I patch in to this configmap created by the ai-rhdh-installer ... I left the other parts for context

gabemontero commented on assets/sidecar/developer-hub-app-config-patches.yaml:
```diff
@@ -0,0 +1,113 @@
+          - host: 'localhost:9090'
+          - host: 'localhost:8080'
+          - host: 'localhost:7070'
+    # end of delta from ai-rhdh-installer
```
start of the changes I patch in to this configmap created by the ai-rhdh-installer ... I left the other parts for context

gabemontero commented on assets/sidecar/developer-hub-app-config-patches.yaml:
```diff
@@ -0,0 +1,113 @@
+---
+apiVersion: v1
+data:
+  RHDH_TOKEN: base64-encoded-token-generated-from-openssl-or-similar-program
```
note ^^ that `RHDH_TOKEN` would need to be set with an actual encoded value for a legitimately built token

gabemontero commented on assets/sidecar/k8s-sa-for-bridge.yaml:
```diff
@@ -0,0 +1,92 @@
+    name: rhdh-rhoai-bridge
+    namespace: ai-rhdh
+---
+# this seemed to make no difference .... I was able to give 'ai-rhdh' permission from their UI ... need to see what roles, etc. that created
```
yeah instead, when I use the model registry UI to add the `ai-rhdh` project per OCP conventions to the list of entities that can access the model registry (per OCP conventions it does not let us add the bridge SA as a User) it creates a dynamically named rolebinging in the `odh-model-registries` NS:

```
gmontero ~/go/src/github.com/redhat-ai-dev/model-catalog-bridge  (sidecar-cfg)$ oc get rolebindings dashboard-permissions-vkwccd -o yaml 
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  creationTimestamp: "2025-03-31T18:38:58Z"
  labels:
    app: modelregistry-public
    app.kubernetes.io/component: model-registry
    app.kubernetes.io/name: modelregistry-public
    app.kubernetes.io/part-of: model-registry
    component: model-registry
    opendatahub.io/dashboard: "true"
    opendatahub.io/rb-project-subject: "true"
  name: dashboard-permissions-vkwccd
  namespace: odh-model-registries
  ownerReferences:
  - apiVersion: modelregistry.opendatahub.io/v1alpha1
    blockOwnerDeletion: false
    kind: ModelRegistry
    name: modelregistry-public
  resourceVersion: "70835"
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: registry-user-modelregistry-public
subjects:
- apiGroup: rbac.authorization.k8s.io
  kind: Group
  name: system:serviceaccounts:ai-rhdh
```

I'll try tomorrow to manually create a non-randomly named rolebinding in a similar fashion and see if it still allows for our SA to query the model registry @thepetk

gabemontero replied:
yeah I was able to create a rolebinding that allows the new SA to query the model registry @thepetk  .... will be part of next push

interestingly, the `ai-rhdh` project is not listed as one of the project in the `modelRegistrySettings/permissions/modelregistry-public` page in the ODH console, but the `rhoai-normalizer` can query the model registry with the rolebinding, and cannot query without the rolebinding

thepetk commented on assets/sidecar/developer-hub-app-config-patches.yaml:
```diff
@@ -0,0 +1,113 @@
+kind: ConfigMap
+metadata:
+  name: developer-hub-app-config
+  namespace: ai-rhdh
```
nit: The namespace should be configurable

gabemontero replied:
yeah I'm assuming you'll make it configurable when you transfer it to your new stuff @thepetk 

I can just remove the setting of the namespace if you like

thepetk commented on pkg/util/k8s-client.go:
```diff
@@ -1,96 +1,168 @@
+          fn := "/opt/app-root/src/dynamic-plugins-root/ca.crt"
+          file, err := os.Create(fn)
+          klog.Infof("create ca.crt file err: %#v", err)
+          if err == nil {
```
nit: we could use early returns with an error here to avoid having too much indentation.

gabemontero replied:
yep good call updated with latest push

thepetk reviewed (commented):
Left a couple of minor comments

thepetk commented:
> ```
> E0331 21:19:41.683594       1 server.go:294] error importing location http://10.128.2.58:9090/mnist/v1/catalog-info.yaml to backstage: Post "http://10.128.2.58:7007/api/catalog/locations": net/http: invalid header field value for "Authorization"
> ```

To clarify, this actually here makes me think that we have access to the url requested, but the value we provide to the `authorization` header is invalid. That said my wild guess for example was if the token has any newline character included.

I would expect something like "invalid host" error trace if we have an issue with the allowed host.

gabemontero commented:
> > ```
> > E0331 21:19:41.683594       1 server.go:294] error importing location http://10.128.2.58:9090/mnist/v1/catalog-info.yaml to backstage: Post "http://10.128.2.58:7007/api/catalog/locations": net/http: invalid header field value for "Authorization"
> > ```
> 
> To clarify, this actually here makes me think that we have access to the url requested, but the value we provide to the `authorization` header is invalid. That said my wild guess for example was if the token has any newline character included.
> 
> I would expect something like "invalid host" error trace if we have an issue with the allowed host.

turns our the token env var somehow has a carriage return / new line when viewed in golang (though I don't see that when I echo it while rsh'ed into the pod) and that was leading to the "bad header format"

once I stripped the newline/crs' I got past that error queries go through

now, that said, I can't get the local pod IP or service IP for the backstage backend to rendor; only using the route's hostname produced content on query

I'll bubble all this up to a more generic sense and then see what happens with the import of a location

thepetk commented:
> turns our the token env var somehow has a carriage return / new line when viewed in golang (though I don't see that when I echo it while rsh'ed into the pod) and that was leading to the "bad header format"

ouch :o  that's very nasty

gabemontero commented:
OK we are not getting as far as we can without having our plugin module running @johnmcollier @thepetk   ... it complains that the location type `rhdh-rhoai-bridge` is not the out of the box location type `url` : 

`post for https://backstage-ai-rh-developer-hub-ai-rhdh.apps.gmontero416.devcluster.openshift.com/api/catalog/locations rc 400 body {"error":{"name":"InputError","message":"Registered locations must be of an allowed type [\"url\"]"},"request":{"method":"POST","url":"/locations"},"response":{"statusCode":400}}`

I'll do some clean up and then push the updates

gabemontero commented:
changes for ^^ pushed @thepetk @johnmcollier 

also, I have TLS enabled in the backstage catalog REST api client running in the `storage-rest` container, since we are accessing the externally accessible route

johnmcollier commented on assets/sidecar/backstage-cr-patches.yaml:
```diff
@@ -0,0 +1,103 @@
+                  - name: STORAGE_TYPE
+                    value: ConfigMap
+                  - name: BKSTG_URL
+                    value: https://backstage-ai-rh-developer-hub-ai-rhdh.apps.gmontero416.devcluster.openshift.com
```
I saw you mention this in a comment, and was curious why neither of the non-route options worked.

Do the sidecars not share the same pod network as the regular Backstage container? And we can't call `localhost:7007`?

gabemontero replied:
was not a k8s networking thing ... both my calls from the golang and curl's when I `oc rsh` into the backend container were able to connect to the various pod or service IP address or address / port combinations for the backstage backend endpoint

they either hung or returned with no data

so it appears to be specific to how RHDH sets up the container/service/route combo ... maybe OCP even does something to protect direct access to the container port ... I would not be surprised, but the OCP networking magic was not where my deepest expertise is.

the simpler bridge endpoints of course work just fine using pod IP .... **and when we get to the backstage extensions trying to contact the location service, aside from configuring the URL in the app-config.yaml @johnmcollier , I suspect you could set the pod IP as a dynamically set env var on the backstage backend container as part of the sidecar setting in this PR, and use that address, and not have to confiure the route in app-config.yaml**

as I am able to make a TLS secure and authenticated connection using the route, I'm no pursuing it any further with this sprint/jira.

certainly with this `BKSTG_URL` config option, any of us are free to experiment with it again at the golang or `oc rsh` level if there is further curiosity to uncover what is going on.

johnmcollier commented on assets/sidecar/developer-hub-app-config-patches.yaml:
```diff
@@ -0,0 +1,113 @@
+              seconds: 15
+            timeout:
+              minutes: 15
+    proxy:
```
Should we remove the non-model catalog specific entries from here?

gabemontero replied:
long term yes (though perhaps in a clean up PR depending on timing)

for now, I'm just trying to give @thepetk clear cut contenxt on where we want to insert the static token and allowed host stuff

johnmcollier commented on assets/sidecar/backstage-cr-patches.yaml:
```diff
@@ -0,0 +1,103 @@
+  name: ai-rh-developer-hub
+  namespace: ai-rhdh
+spec:
+  application:
```
Do we need to reference the app config you created below?

gabemontero replied:
@johnmcollier and @thepetk and I talked and I now understand that @johnmcollier thinks it better to have a full of the backstage CR in case a developer does not use the installer 

Ultimately that makes our dependencies explicit and clear

I will make the update as part of my next push

johnmcollier commented on assets/sidecar/developer-hub-app-config-patches.yaml:
```diff
@@ -0,0 +1,113 @@
+
+---
+apiVersion: v1
+data:
```
Is this configmap for testing, or part of what will be deployed?

IMO - I think it's fine for testing, but for production use cases, we'll probably want to make sure the fields we're adding merge fine with the user's pre-existing app-config.yamls (if there's any overlap)

gabemontero replied:
yeah I consider this whole `sidecar` folder for dev testing, and never beyond that, and a reference starting point for @thepetk to use gitops, kustomize, etc. for the production use case 

I thought from one of the recent video conf's @thepetk was fine with that

I should probably update the README with something to that affect i.e. it is just a ref for dev testing, and assumes for now you've used the ai-rhdh-installer

that said, once @thepetk 's work is done for the rolling demo, I am more inclined to have this repo simply point to that for setting up the side car, stand alone bridge pod, etc.

WDYT @thepetk @johnmcollier

johnmcollier commented on assets/sidecar/k8s-sa-for-bridge.yaml:
```diff
@@ -0,0 +1,92 @@
+---
+apiVersion: v1
```
Not related to the serviceaccount, but one of your comments got me thinking:

> Maybe that means my allowed host setting of 10.128.* is not valid for allowing the import, and we have to define a service ? ....

(I couldn't tell if this resolved or not from later comments, but) IMO, it's probably worth creating a service just so that we can hard code the service name and reference that instead of some IP.

gabemontero replied:
I think it is resolved , in that `'10.*:9090'` seems to be sufficiently generic, effectively hard coded, yet exclusive, but the proof will truly be once the backstage extensions are loaded into RHDH and we can actually push new models to the model catalog

also, remember, the service name host name will have the namespace in it, so it is not truly fully hard coded, depending on where RHDH is deployed

using the dynamically inserted POD IP address env var I'm suggesting the entity provider uses to contact the location service will not have this namespace variable 

we'll certainly need services long term when it runs in a standalone pod, but otherwise, for the sidecar, I don't think it buys us anything assuming I'm correct about `10.*:9090` for the allowed host.  K8s pod IPs always start with 10

gabemontero replied:
@johnmcollier and I sorted on our call we are good here

johnmcollier reviewed (commented):
Just a couple comments, nice work around you found

gabemontero commented:
ok @johnmcollier @thepetk I have rebased with the json array changes merged, and I've posted responses to @johnmcollier 's comments, though I have not resolved those comments, and more discussion seems conceivable

gabemontero commented:
@johnmcollier @thepetk I've rebased this PR to include the changes https://github.com/redhat-ai-dev/model-catalog-bridge/pull/16 that @johnmcollier and I worked out during his plugin dev today

I updated the README with the non-trivial steps needed to manually update the various RHDH config set up from the ai-rhdh-installer to get the sidecars running .... summary, use `oc edit` cause `oc patch` is too onerous

lastly, guessing @johnmcollier is not far off, I updated the format setting to the new json array format, and validated with local testing it shows up in the storage-rest layer's configmap in json format

gabemontero commented:
OK @johnmcollier @thepetk per our call today I have 
- updated the post ai-rhdh-installer yaml in the `assets/sidecar-after-ai-rhdh-installer` folder to be a full snapshot
- I've added comments about what are the deltas, and what secret/config reference names in the `Backstage` instance are dynamically generated and need to be refreshed if you choose to do a full replacement
- per a post mortem with @johnmcollier the deltas for creating requisite yaml that could be applied on a vanilla/default instance from the operator were too onerous .... when you consider non-guest auth, templates that use the models, etc. the minimal dev/demo env is non trivial ... we'll potentially iterate on this after summit, though the helm based gitops repo could prove to be a sufficient "dev env" longer term
- I've supplied to @thepetk in https://redhat-internal.slack.com/archives/C07N7CR4FFH/p1744133331048139?thread_ts=1744039487.693989&cid=C07N7CR4FFH  the deployment / configmap / secret yamls to go with the new `k8s-sa-for-bridge.yaml` created in this PR that can be factored into his non-operator based install for the rolling demo

Per our call merging this PR

gabemontero commented:
https://github.com/redhat-ai-dev/model-catalog-bridge/pull/18 will update the backstage CR yaml with the images pushed to quay as a result of this PR merging
